### PR TITLE
PUD-1654  Weekly recalls new report Back end: initial thread.

### PR DESCRIPTION
**Pull request by AndyNewton**
PUD-1654  Weekly recalls new report Back end: initial thread.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/ReportsController.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@RequestMapping(produces = [MediaType.APPLICATION_JSON_VALUE])
+@PreAuthorize("hasRole('ROLE_MANAGE_RECALLS')")
+class ReportsController(
+  @Autowired private val reportsService: ReportsService,
+  @Autowired private val clock: Clock
+) {
+
+  @GetMapping("/reports/weekly-recalls-new")
+  fun weeklyRecallsNew(): GetReportResponse {
+    val now = OffsetDateTime.now(clock)
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/ReportsController.kt:24]: feels odd to have the date here, but 🤷‍♂️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/ReportsService.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  private val headers = arrayOf(
+    "CUSTODY_TYPE_DESCRIPTION",
+    "CUSTODY_TYPE_AT_TIME_OF_RECALL_DESCRIPTION",
+    "FAMILY_NAME",
+    "FIRST_NAMES",
+    "NOMS_ID",
+    "PRISON_NUMBER",
+/*
+GENDER - n/a today; load in future?
+LICENCE_REVOKE_DATE - not stored on recall; save in future?
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/ReportsService.kt:28]: I think we have this, its the date that the RecallNotification is sent I believe

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt [hunk 3]
```diff
@@ -96,22 +96,8 @@ class ManagerRecallsUiAuthorizedPactTest : ManagerRecallsUiPactTestBase() {
...
-    phaseRecordRepository.deleteAll()
-    partBRecordRepository.deleteAll()
-    lastKnownAddressRepository.deleteAll()
-    rescindRecordRepository.deleteAll()
-    noteRepository.deleteAll()
-    documentRepository.deleteAll()
-    recallRepository.deleteAll()
-
+  fun `delete all recalls and set no virus found`() {
+    `delete all recalls`()
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/pact/provider/ManageRecallsUIPactTest.kt:100]: maybe a better name? Its delete all data?

**Review by philsegal — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/ReportsController.kt (hunk 1 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/controller/ReportsController.kt:24]: The idea was to use the clock as early as possible and as few times as possible - as it is only present to support e.g. TZ config via Spring and testing.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/RecallRepository.kt [hunk 4]
```diff
@@ -18,6 +18,8 @@ import java.util.UUID
 @Repository("jpaRecallRepository")
 interface JpaRecallRepository : JpaRepository<Recall, UUID> {
   fun findAllByNomsNumber(nomsNumber: NomsNumber): List<Recall>
+
+  fun findAllByCreatedDateTimeIsBetweenOrderByCreatedDateTimeAsc(startCreatedDateTime: OffsetDateTime, endCreatedDateTime: OffsetDateTime): List<Recall>
```
[the remark this answers is not in the document]
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/RecallRepository.kt:22]: I guess we'll find out!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/ReportsService.kt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    val data = recallRepository.findAllByCreatedDateTimeIsBetweenOrderByCreatedDateTimeAsc(start, end)
+    val out = StringBuilder()
+    val format = CSVFormat.DEFAULT.withHeader(*headers)
+    CSVPrinter(
+      out, format
+    ).use { printer ->
+      data.forEach { r ->
+        printer.printRecord(recallCsvRecord(r))
+      }
+    }
```
[the remark this answers is not in the document]
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/ReportsService.kt:50]: Yes; this `.use {}` pattern is less than obvious, I agree.

**Review by jonnywyatt — APPROVED** [review]

**Pull request merged**